README.md
# Flutter - Bubble Animated Tabbar

### If this project has helped you out, please support us with a star. :star2:

## Preview

![Alt Text](tab.gif)

## Page View

![Alt Text](page-1.gif)

<style type="text/css">
    img{ border:1px solid black }
</style>

## Getting Started

```yamlgit
dependencies:
	...
	bubble_animated_tabbar: 1.0.2
```

## Usage Example

Adding the widget

```dart
// Array of Tabs

List<Map> children = [
    {
      'icon': Icons.home,
      'title': 'Home',
      'iconSize':23,
      'color': Color.fromRGBO(186, 121, 9, 0.25),
      'textColor': Color.fromRGBO(173, 118, 23, 1),
      'customTextStyle':TextStyle(color: Color.fromRGBO(173, 118, 23, 1)),
    },
    {
      'icon': Icons.mail,
      'title': 'Messages',
      'color': Color.fromRGBO(0, 2, 148, 0.22),
      'textColor': Color.fromRGBO(26, 1, 135, 0.9),
    },
    {
      'icon': Icons.person,
      'title': 'User Account',
      'color': Color.fromRGBO(18, 140, 126, 0.25),
      'textColor': Color.fromRGBO(6, 125, 111, 1),
    },
  ];

  getBoxDecoration() {
    return BoxDecoration(
      color: Colors.white,
      boxShadow: [
        new BoxShadow(
          color: Color.fromRGBO(0, 0, 0, 0.1),
            blurRadius: 1.0,
            spreadRadius: 1.0
          )
        ]
    );
  }

// Add required properties for Animated TabBar
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar( title: Text(children[_currentIndex]['title']) ),
      bottomNavigationBar: AnimatedTabbar(
        padding:EdgeInsets.only(left: 4, top: 6, right: 4, bottom: 6),
        containerDecoration: getBoxDecoration(),
        currentIndex: _currentIndex,
        onTap: onTabTapped,
        children: children,
      ),
    );
  }
```

## AnimatedTabbar Properties

| Property            | Type          | Required | Default Value |
| ------------------- | ------------- | -------- | ------------- |
| containerDecoration | BoxDecoration | NO       | --            |
| padding             | EdgeInsets    | NO       | --            |
| onTap               | Function      | Yes      | --            |
| children            | Array < Map > | Yes      | --            |

## Tab Properties

| Property        | Type       | Required | Default Value                     |
| --------------- | ---------- | -------- | --------------------------------- |
| icon            | IconData   | YES      | --                                |
| iconSize        | Int        | NO       | 20                                |
| title           | String     | YES      | --                                |
| color           | Color      | YES      | --                                |
| textColor       | Color      | YES      | --                                |
| tabPadding      | EdgeInsets | NO       | EdgeInsets.fromLTRB(10, 4, 10, 4) |
| customTextStyle | TextStyle  | NO       | --                                |


CHANGELOG.md
## [0.0.1] - TODO: Add release date.
* TODO: Describe initial release.

## [1.0.1] 
* Updated Readme.md file

test/bubble_animated_tabbar_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:bubble_animated_tabbar/bubble_animated_tabbar.dart';

void main() {}


lib/bubble_animated_tabbar.dart
library bubble_animated_tabbar;

import 'package:flutter/material.dart';

class AnimatedTabbar extends StatefulWidget {
  final int currentIndex;
  final onTap;
  final BoxDecoration containerDecoration;
  final EdgeInsets padding;
  List<Widget> tabBarChildren = [];

  AnimatedTabbar({Key key, @required this.currentIndex, @required this.onTap, @required List<Map> children, @required this.padding, this.containerDecoration})
      : super(key: key) {
    children.asMap().forEach((int index, Map tabData) {
      tabBarChildren.add(
        Expanded(
          flex:0,
          child: AnimatedTabBarItem(
            tabData: tabData, index: index, onTap: onTap, tabState: getTabState(children, tabData)
          )
        )
      );
    });
  }

  getTabState(children, data) {
    return children.indexOf(data) == this.currentIndex ? "ACTIVE" : "INACTIVE";
  }

  @override
  AnimatedTabbarState createState() => AnimatedTabbarState();
}

class AnimatedTabbarState extends State<AnimatedTabbar> {

  getContainerDecoration(){
    return widget.containerDecoration ?? BoxDecoration();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      // color:Color.fromRGBO(0,0,0,.21),
      decoration: getContainerDecoration(),
      padding: widget.padding,
      child: Row(
        children: widget.tabBarChildren,       
        mainAxisAlignment: MainAxisAlignment.spaceAround, 
        crossAxisAlignment: CrossAxisAlignment.center        
      )
    );
  }
}

class AnimatedTabBarItem extends StatefulWidget {
  final tabData;
  final String tabState;
  final onTap;
  final int index;

  void onTabTap() {
    onTap(index);
  }

  AnimatedTabBarItem(
    {
      Key key, 
      @required this.tabData, 
      @required this.tabState, 
      @required this.onTap, 
      @required this.index
    }): super(key: key);

  @override
  AnimatedTabBarItemState createState() => AnimatedTabBarItemState();
}

class AnimatedTabBarItemState extends State<AnimatedTabBarItem> with SingleTickerProviderStateMixin {
AnimationController controller;

@override
  void initState() {
    super.initState();
  }

  @override
	void dispose(){
	  super.dispose();
	}

  getTabColor(){
    return widget.tabData['color'];
  }

  getTabFontColor(){
    return widget.tabState == "ACTIVE" ? widget.tabData['textColor'] : widget.tabData['inactiveIconColor'] ?? Color.fromRGBO(0,0,0,0.65);
  }

  Text getTabTextWidget(){
    if(widget.tabState == "ACTIVE"){
      if(widget.tabData['customTextStyle'] !=null){
        return new Text(widget.tabData['title'] ?? "", style: widget.tabData['customTextStyle']);
      }else{
        return new Text(widget.tabData['title'] ?? "", style: TextStyle(color: getTabFontColor()));
      }
    } else{
      return Text("");
    }
  }

  wrapAnimatedContainer(String tabState){
     return AnimatedSize(
      vsync: this,
      curve: Curves.fastOutSlowIn,
      duration: Duration(milliseconds: 250),
      reverseDuration: Duration(milliseconds: 150),
      child: Container(
        padding: tabState == "ACTIVE" ? EdgeInsets.only(left: 6, right: 2):EdgeInsets.all(0), 
        child: getTabTextWidget()
      )
    );
  }

  @override
  Widget build(BuildContext context) {
    Widget element;
      element = GestureDetector(
        onTap: widget.onTabTap, 
        child: AnimatedContainer(
          duration:Duration(milliseconds: 120),
          curve:Curves.easeInCirc,
          padding: widget.tabData['tabPadding'] ?? EdgeInsets.fromLTRB(10, 4, 10, 4),          
          decoration: BoxDecoration(
            borderRadius: BorderRadius.all(Radius.circular(30)),
            color: widget.tabState == "ACTIVE" ? getTabColor():Color.fromRGBO(0,0,0,0),
          ),
          child:
            Row(
              children: [
                new Icon(widget.tabData['icon'], color: getTabFontColor(), size: (widget.tabData['iconSize']?? 20).toDouble()),
                wrapAnimatedContainer(widget.tabState)                
              ]
            )
        ) 
      );
    return SafeArea(child: element);
  }
}


example/lib/home.dart
import 'package:flutter/material.dart';
import 'package:bubble_animated_tabbar/bubble_animated_tabbar.dart';

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with SingleTickerProviderStateMixin {
  int _currentIndex = 0;
  TabController tabController;
  List<Map> children = [
    {
      'icon': Icons.home,
      'title': 'Home',
      'iconSize':23,
      'color': Color.fromRGBO(186, 121, 9, 0.25),
      'textColor': Color.fromRGBO(173, 118, 23, 1),
      'customTextStyle':TextStyle(fontSize: 17, color: Color.fromRGBO(173, 118, 23, 1)),
      'tabPadding': EdgeInsets.fromLTRB(10, 6, 10, 6)
    },
    {
      'icon': Icons.mail,
      'title': 'Messages',
      'iconSize':23,
      'color': Color.fromRGBO(0, 2, 148, 0.22),
      'textColor': Color.fromRGBO(26, 1, 135, 0.9),
      'customTextStyle':TextStyle(fontSize: 17, color: Color.fromRGBO(26, 1, 135, 0.9)),
      'tabPadding': EdgeInsets.fromLTRB(10, 6, 10, 6)
    },
    {
      'icon': Icons.person,
      'title': 'User Account',
      'iconSize':23,
      'color': Color.fromRGBO(18, 140, 126, 0.25),
      'textColor': Color.fromRGBO(6, 125, 111, 1),
      'customTextStyle':TextStyle(fontSize: 17, color: Color.fromRGBO(6, 125, 111, 1)),    
      'tabPadding': EdgeInsets.fromLTRB(10, 6, 10, 6)
    },
    {
      'icon': Icons.settings,
      'title': 'Settings',
      'iconSize':23,
      'color': Color.fromRGBO(96, 0, 148, 0.25),
      'textColor': Color.fromRGBO(96, 0, 148, 1),
      'customTextStyle':TextStyle(fontSize: 17, color: Color.fromRGBO(96, 0, 148, 1)),
      'tabPadding': EdgeInsets.fromLTRB(10, 6, 10, 6),
    },
  ];

  @override
  void initState() {
    super.initState();
    tabController = TabController(vsync: this, initialIndex: 0, length: children.length);
    tabController.addListener(onTabViewChange);
  }

  onTabViewChange() {
    setState(() {
      _currentIndex = tabController.index;
    });
  }

  getBoxDecoration() {
    return BoxDecoration(
        color: Colors.white,
        boxShadow: [new BoxShadow(color: Color.fromRGBO(0, 0, 0, 0.1), blurRadius: 1.0, spreadRadius: 1.0)]);
  }

  void onTabTapped(int index) {
    setState(() {
      _currentIndex = index;
      tabController.animateTo(index);
    });
  }

  getPage() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Text('You are on Page'),
          Text('${_currentIndex + 1}',style: Theme.of(context).textTheme.display1),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(children[_currentIndex]['title']),
      ),
      body: TabBarView(
        children: [getPage(), getPage(), getPage(), getPage()],
        controller: tabController,
      ),
      bottomNavigationBar: AnimatedTabbar(
        padding:EdgeInsets.only(left: 4, top: 6, right: 4, bottom: 6),
        containerDecoration: getBoxDecoration(),
        currentIndex: _currentIndex,
        onTap: onTabTapped,
        children: children,
      ),
    );
  }
}
